### DISCUSSION TITLE
Optimize for recognizing more barcodes?

### DISCUSSION THREAD

**Discussion by duanqn** [zxing-cpp/zxing-cpp#398] [Q&A] [answered]
Hi,
If I want to recognize as many barcodes as possible, and I don't care speed, what should be the hints I pass to ZXing-cpp?
Also my images all contain 1 barcode (but the orientation can be 0 / 90 / 180 / 270). I'm looking for some suggestions for the recognition options (`DecoderHints`). Thanks!

**Comment by axxel** [comment, marked as answer]
The default values are already chosen such that the detection rate is maximized. If you care about DataMatrix codes, you should compile the library in c++-20 mode (see #189).

One way to potentially increase the detection rate slightly is to call `ReadBarcode` twice, once with the `LocalAverage` binarizer and then again with the `GlobalHistogram` one. This only makes a difference with matrix symbologies, tough, not with linear ones.

**Reply by axxel** [reply]
I forgot to mention `downscaleThreshold` and `downscaleFactor`. Lowering them from their defaults (500 and 3) might help in some corner cases (but it seems you were aware of that already...).

**Reply by duanqn** [reply]
Thanks!

**Comment by Sergio-** [comment]
Hi Folks. Good day. I hope you're doing good.
I'm researching methods to improve the decoding of high density QR Codes that have some level of damage or low contrast.
I'd like to know if it is possible to apply some type of pre-processing to the image (brightness/contrast, blur/sharpen, threshold, etc) before sending the image to the decoding algorithm.
I've attached some examples where the Binary Eye Android App struggles to decode, but a standard/simple usb reader can decode them the first second.
Bottom line: Is there any methods you guys can think of to improve the decoding on these type of situations?
Appreciate any feedback/suggestion/collaboration.
![image](https://user-images.githubusercontent.com/3593635/208824833-485a7638-acd1-48ab-a877-f41a9c1e0395.png)
![image](https://user-images.githubusercontent.com/3593635/208824848-375ee917-51fb-4804-a4ef-7b504ba0073e.png)
![image](https://user-images.githubusercontent.com/3593635/208824857-774de1a6-5aa4-4242-842c-d297fef74a2e.png)

**Reply by axxel** [reply]
I had a look at all of your 3 samples and can provide the following insights:
1. the 'only' problem with the first image is the broken upper/right finder pattern. If I make the 'shortened' 3x3 center square a little wider (paint over the missing black part), it is detected as is.
2. the second image also has a broken (upper/left) finder pattern, that prevents the detection. But if I fix that manual, I still get a checksum error, because there are too many washed out modules.
3. the third can be fixed by applying an erosion filter one time.

Bottom line: your samples are challenging. They have some 'style' in common where I would believe you can improve the overall detection by playing with something like erosion/dilation or different thresholding approaches that are tuned for these inputs.

There is also an opportunity to make some more intrusive changes to the QRDetector and the whole grid sampler to incorporate the intermediate alignment patterns, which are currently completely ignored. But those improvements alone will not make these 3 samples work as the finder patterns still need to be sufficiently intact for the symbol to be picked up at all.

**Comment by Sergio-** [comment]
HI @axxel Good day. I hope you're doing good.
Thanks for the reply on this topic.
In parallel, I've posted a similar question on BinaryEye repository (as it is the App I'm using for testing/comparison). The thread contains more details, its here: https://github.com/markusfisch/BinaryEye/issues/337
It seems it makes more sense to discuss possible improvements on the ZXing-Cpp library itself (in opposite of the App). So, I'll continue the discussion here.
I've conducted more tests. I've recorded another video using another scanned qrcode that, can be decoded with Dynamsoft online demo (and on a USB reader) but not in BinaryEye (v1.57.0) using ZXing-CPP (v1.4.0.11). See below.

https://user-images.githubusercontent.com/3593635/210263933-52a632fa-9e42-469a-95ff-f8963d7ce5c3.mp4


Here's the QRCode used.
![image](https://user-images.githubusercontent.com/3593635/210263848-3cfcfa17-7d48-4bd1-a7b1-f349a480305f.png)

I've applied some basic processing using GIMP on that Image, and, with that, we can dramatically improve the performance, with images being decoded. See below.

![image](https://user-images.githubusercontent.com/3593635/210265523-ca6a9b69-edbf-4d76-a9b8-bb4388fac3a0.png)


I did several other tests with similar challenging images, in most of cases, the Color Level + Eroding (once or twice) provide significant improvements.

The question is: Do you see value in implementing some of these basic filtering in the library? Does it make sense to have it integrate in the library or should it be external to it?

Although I'm not skilled to implement it myself, I can help in other areas like testing if needed.

What are your thoughts on it?

Some additional information: I've processed that image a bit more to achieve a best contrast. See below:
![image](https://user-images.githubusercontent.com/3593635/210269043-e58d3aa1-232c-4535-95c8-cc7e4a05b47e.png)

That's a interesting case, The image seems to be fine (not major damage), It can be decoded very easily/fast on the online proprietary library demo and it also reads very well on a USB reader (reading directly from screen or printed on a paper), but BinaryEye doesn't seems to be able to decode it. I'm not sure why is that the case... What is missing on that image and what is the "secret sauce" that makes the proprietary library and the USB reader (with a 640x480 camera) read/decode it so easily?

https://user-images.githubusercontent.com/3593635/210269349-dc899648-2577-46cc-a0c9-3e0d35f6f616.mp4


Obs: The tests were conducted on Samsung Galaxy Note 20 Ultra with stock Android 12.

FYI @markusfisch 

Thanks a lot again!

**Reply by axxel** [reply]
Observations about your last sample:
 * contrast is totally sufficient
 * there is a slight geometric distortion causing this image not being detected if put directly into lib via `ZXingReader`
 * if scanned from the screen with my phone (like in your videos) both the zxing-cpp demo Android app and BinaryEye can detect in on my Google Pixel
 * the demo-app detects it considerably faster/earlier. maybe because of the `tryHarder` or the `tryDownscale` flag not being used inside BinaryEye
 * the secret source in this case might be a proper analysis of the alignment patterns inside the symbol. there is no trivial fix for that, though.

**Comment by Sergio-** [comment]
HI @axxel Thanks for the reply.
My ultimate goal when starting this discussion is to see what can be done in terms of improving the decoding accuracy for these "non optimal" cases.
1) Do you see any feasible improvement (like implementing a contrast/erode for example) on the library so it can decode some patterns easily?
2) Is there any other way you can think of to optimize/improve the decoding of these images?
I don't know the inner workings of the library... I guess some of my questions are how a USB barcode reader with a low resolution camera can decode these images and a mobile camera with a high res cannot? Well, maybe it's question of the algorithm optimization as with the same camera, that proprietary library have a very high accuracy... At this points, I'm wondering if we can expect similar (or better :-) ) performance on ZXing library by implementing some pre-processing filters... 

Thanks a lot!

**Reply by axxel** [reply]
1) I don't think implementing some extra contrast/erosion algorithms are going to help in general. What you did to make this image work would break others. The difficulty here is what you did as a human, i.e. already recognizing what the barcode is and then tuning until it looks good.
2) Yes there is, the mentioned proper analysis of the alignment patterns inside the symbol. The question is: who has the skills and the time to do it. "some pre-processing filters" will not make zxing-cpp become as performant as the commercial library you compared it to.

**Comment by Sergio-** [comment]
Hi @axxel Thanks for the feedback. I guess it is what it is then. But thanks for clarifying it.

I know it's very specific, but, the use case I'm researching only will make use of that particular scenario, so, I'm fine if, for my application, I can tune the library to that scenario, (even if it breaks the other decoding)... 
Do you have any little guidance or tip to where to look into in the source code to experiment with that? 
Any help is appreciate. If there's any suggestion on what I can start experimenting, I appreciate it.
Best Regards!

**Reply by axxel** [reply]
If your application is (only) dealing with similar input as the one above then coming up with some erosion/contrast preprocessing filter will help you. You can do that with whatever hand-written or external library (I would suggest to look at opencv) based code outside of zxing-cpp. If you also have a rough idea where your barcode is (like in the hand-held phone scenario above), crop your image accordingly before passing it in and also reduce the resolution such that a single module (the smallest black little square) is not more than around 8 pixels wide.

The alignment pattern detection and sample grid adjustment I mentioned, would help all codes. If you want to look into that, start [here](https://github.com/zxing-cpp/zxing-cpp/blob/f1f3494bbeccc6d2f5b16b110fa3ed955871f317/core/src/qrcode/QRDetector.cpp#L260)

[If you want to reply to this message, please use the "Write a reply" field directly below this text instead of adding a new answer to the top-level discussion]

**Comment by Sergio-** [comment]
Hi @axxel Thanks for the response.
I'll look at the it. Thank you very much.
Also, As you are the main maintainer of this library, are you open/accepting any funding that might improve help with the current roadmap? Best Regards.

**Reply by axxel** [reply]
> Also, As you are the main maintainer of this library, are you open/accepting any funding that might improve help with the current roadmap?

I'm certainly open to the idea :). I did in-fact once drop a very limited [test-balloon comment](https://github.com/zxing-cpp/zxing-cpp/issues/354#issuecomment-1175959225) in this direction a couple months back. I'm self-employed with my business still being affected by the pandemic aftermath, which is the reason why I could spend so much time in the last 2 years on this "hobby" project. In what way would you be interested in contributing funding, if I may ask?

[Again: to make the whole discussion more readable, please answer in the "Write a reply" text field right below this line, not the "Suggest an answer" field at the bottom of the page. Keeps this together as a single thread.]

**Reply by Sergio-** [reply]
Hi @axxel. Let's talk a bit about it. 
If possible, please, mail me (sergio [dot] olivo [at] gmail). And we can setup a call if you agree.
Best Regards and apologies for posting the responses in the wrong field :-)
